### ISSUE TITLE
Scans interface drops IMU packets

### ISSUE THREAD

**Issue by kpanyam** [ouster-lidar/ouster_example#455]
I am trying to rewrite our custom C code to read Lidar and IMU packets from Ouster, using the Python SDK 0.5.1. Ideally I would like to use the Scans interface and let Ouster deal with combining the Lidar packets info frames.

The problem is that the Scans class simply ignores IMU packets. See the code snippet from _ouster/client/core.py_. There is no else clause for if _isinstance(packet, LidarPacket)_. It should handle the IMU packets in the else clause, but then it breaks the scheme that this class returns LidarScan only. 

It would be great if you fix this (or add another interface) to return both Lidar scans as well as IMU packets.

```
class Scans:
    def __iter__(self) -> Iterator[LidarScan]:
        """Get an iterator."""
        ...
        while True:
            try: 
                packet = next(it) 
            except StopIteration:
            ...
            if isinstance(packet, LidarPacket): 
                ...
```

**Comment by bexcite** [comment]
Hi @kpanyam,

Yes it's a great point and we have some work to do to provide nice easy to use wrappers for such interfaces, meanwhile you can use a wrapper modeled below `LidarDataSource` that returns batched `LidarScan` and `ImuPacket` with a corresponding timestamp of the UDP packets. It can work easily with `Pcap` source or `Sensor` source:
```
from ouster import client
import ouster.pcap as pcap

import numpy as np
from typing import cast, Iterator, Union, Tuple, Dict, Optional, List


# (ts, ImuPacket) or (ts, LidarScan), ts in nanoseconds
LidarData = Union[Tuple[int, client.ImuPacket], Tuple[int, client.LidarScan]]

class LidarDataSource:
    """Make a packet source with collected LidarScans and preserved ImuPackets
    so one can easily work with the stream of final data objects from Pcap files
    and avoid double read (for imu/lidar packets) or more complex iterator
    wrappings.
    """

    def __init__(
        self,
        source: client.PacketSource,
        *,
        complete: bool = False
    ) -> None:
        self._source = source
        self._complete = complete
        self._fields = self._source.metadata.format.udp_profile_lidar

    def __iter__(self) -> Iterator[LidarData]:
        w = self._source.metadata.format.columns_per_frame
        h = self._source.metadata.format.pixels_per_column

        ls_write_started = False
        ls_write_start_ts = 0
        ls_write = client._client.LidarScan(h, w, self._fields)

        pf = client._client.PacketFormat.from_info(self._source.metadata)
        batch = client._client.ScanBatcher(w, pf)

        for packet in self._source:
            ts = int(packet.capture_timestamp * 10**9) if packet.capture_timestamp else 0
            if isinstance(packet, client.ImuPacket):
                yield ts, packet
            elif isinstance(packet, client.LidarPacket):
                if not ls_write_started:
                    ls_write_start_ts = ts
                    ls_write_started = True
                if batch(packet._data, ls_write):
                    if not self._complete or ls_write.complete():
                        yield ls_write_start_ts, ls_write
                    ls_write = client._client.LidarScan(h, w, self._fields)
                    ls_write_start_ts = ts

    @property
    def metadata(self) -> client.SensorInfo:
        return self._source.metadata


# Example LidarDataSource usage
source: client.PacketSource
source = .... # pcap.Pcap(file, info) OR client.Sensor(hostname, lidar_port, imu_port)

for ts, obj in LidarDataSource(source):
    if isinstance(obj, client.LidarScan):
        ls: client.LidarScan
        ls = obj
        # process ls as LidarScan with a ts as the timestamp of scan beginning (in HOST time)
        # ....
    elif isinstance(obj, client.ImuPacket):
        imu: client.ImuPacket
        imu = obj
        # process imu as ImuPacket  with a ts as the timestamp of imu packet (in HOST time)
        # ...
```

The original design decision was to have a `PacketSource` that is general enough and can be used to produce other specialized streams, and `Scans` is that super specialized stream that returns only `LidarScan` because not all users care about `ImuPacckets` and it's simpler for them to just not even have additional checks for `isinstance()`. 

To collate further `ImuPackets` that goes together in time with `LidarScan` you can make a further wrapper and decide what time you want to use for this, because there are two times that you may consider:
1) UDP packets time (i.e. HOST time) (used in example)
2) LidarScan `LidarScan.timestamp[]` or `ImuPacket.sys_ts()` one can further get from `ls` and `imu` objects. (i.e. SENSOR time)

Hope it helps.

Cheers,

**Comment by kpanyam** [comment]
Thanks. I will give this a try. 

Could you also tell me how to cleanly get the metadata bytes (to save and use offline to read stored data)? I am currently using `<source>.write_metadata(<filename>)` and reading the json file to get the bytes, but this is not so good.

**Comment by bexcite** [comment]
Yes, getting the json metadata from the `PacketSource` is not possible right now (we have a task to make it all better in the future), however if you use the `Sensor()` and you are not providing the `metadata` in the ctor, you can use the internal `Sensor` member `_fetched_metadata` to access to the json as it was fetched from the sensor: https://github.com/ouster-lidar/ouster_example/blob/master/python/src/ouster/client/core.py#L159

WARNING: Because it's an internal property it can be changed easily in the next releases, so use at your own risk.

**Comment by kairenw** [comment]
What is `source` though? if it's `Sensor` class, then `<source>.write_metadata` _is_ the supported way

**Comment by kpanyam** [comment]
I am already doing write_metadata() and then reading the saved file, but I
would like a more elegant way to do this. I didd see _fetched_metadata, but
then I don't want to use undocumented proprties.

**Comment by kairenw** [comment]
Can you go into why it's inelegant/not great for you?

**Comment by kpanyam** [comment]
We record lidar frames in our own custom format, and I would like to write
the metadata bytes into it as a header, for reading offline. So I am
looking for a way to get the metadata bytes. I don't really care to write a
file just to read it immediately.

You can tell me which one is more elegant.

fname = "/tmp/metadata.json"
source.write_metadata(fname)
with open(fname) as f:
      metadata = bytes(f.read(), "utf-8")
os.remove(fname)

or something like

metadata = source.get_metadata....()

**Comment by kairenw** [comment]
yes, i see, i'd use `_fetched_meta` as @bexcite suggested

**Comment by kairenw** [comment]
I've added the enhancement label to this issue. We will keep it open until we can address it, but no promises on timeline
